### ISSUE TITLE
Different results with whisper and whisper_timestamped

### ISSUE THREAD

**Issue by skanda1005** [linto-ai/whisper-timestamped#38]
When I run an audio file through whisper I get good transcription but when it is run through whisper_timestamped the results are different with the same parameters.
Audio file:
https://drive.google.com/file/d/11FefP13yHkKOgSMUDjwRR3ztbtAfZCSD/view?usp=sharing

whisper:
script = model.transcribe(audio, language = 'Hindi', beam_size = 5, best_of=5, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0))

whisper_timestamped:
>>> audio = whisper.load_audio('/data_drive_500/vocals_skanda_sil.wav')
>>> result = whisper.transcribe(model, audio, language='hi', beam_size=5, best_of=5, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0))

**Comment by Jeronymous** [comment]
The thing is that whipser is not deterministic, when the sampling temperature > 0.
If you run it twice, you should also see different results.

Because I don't like this random behavior (lack of reproducibility), I already added a random seed in whisper_timestamped, which will always give you the same result.

You should get the same results between whisper and whisper_timestamped by adding, before running whisper:
```
import torch
seed = 1234
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
```

If you can't get the same results, then tell me and I will investigate more (just please specify the model you use).

Note: you don't say in your description if the results are also good with whisper_timestamped.
I would be surprised if there would be significantly worse (if it's just a matter of random seed)

**Comment by Jeronymous** [comment]
@skanda1005 did you get a chance to
- run whisper several times and observe that you also get different results (independently of whisper_timestamped)
- check that you get the same results as whisper_timestamped by setting seed 1234 before calling whisper?

(this may clarify my previous message)

Note: This holds when some temperature > 0 are involved in the decoding process.
Which, by the way, does not necessarily happen when the first temperature to try is 0 (`temperature=(0.0, 0.2, ...)`). It depends on the audio, the model...

**Comment by skanda1005** [comment]
> @skanda1005 did you get a chance to
> 
> * run whisper several times and observe that you also get different results (independently of whisper_timestamped)
> * check that you get the same results as whisper_timestamped by setting seed 1234 before calling whisper?
> 
> (this may clarify my previous message)
> 
> Note: This holds when some temperature > 0 are involved in the decoding process. Which, by the way, does not necessarily happen when the first temperature to try is 0 (`temperature=(0.0, 0.2, ...)`). It depends on the audio, the model...

Yes, I have noticed that whisper gives new results every time I run and answering your previous question, I wouldn't say whisper_timestamped is giving too bad a result, but it is definitely skipping more sentences than whisper. Will check it and see with. a seed of 1234 and confirm the results

**Comment by macnibblet** [comment]
@skanda1005 Did you see this section in the readme?

> Besides, default decoding options are different, in order to favour efficient decoding (greedy decoding instead of beam search, and no temperature sampling fallback). To have same default as in whisper, use beam_size=5, best_of=5, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0).

I missed it the first time while going through the docs and changing it back to the default whisper settings gives me better results.

**Comment by Jeronymous** [comment]
From the description of his issue, it seems that @skanda1005 is doing this correctly:
> > > result = whisper.transcribe(model, audio, language='hi', beam_size=5, best_of=5, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0))

**Comment by Rtut654** [comment]
@Jeronymous It might be an off topic, but I'm getting different results on whisper and whisper-timestamped. Both seems incorrect in terms of locating mispronunciation. 
Audio to transcribe : https://drive.google.com/file/d/12zGWllJg6edftcnwuHX_ZHMuwk7PlVjg/view?usp=sharing

Code for whisper-timestamped:
```
import whisper_timestamped 
from colorama import init, Style
from typing import List
import colorsys

AUDIO = 'audio/audio_2023-08-16_12-22-08.wav'

audio = whisper_timestamped.load_audio(AUDIO)
model = whisper_timestamped.load_model("small", device="cuda:4")
results = whisper_timestamped.transcribe(model, audio=audio,beam_size=5, language='en')

output_text = ''
init(autoreset=False)
for segment in results['segments']:    
    for word in segment['words']:
        text = word['text']
        confidence = word['confidence']
        
        r, g, b = colorsys.hsv_to_rgb(confidence * (1/3), 1, 1)
        r, g, b = int(r * 255), int(g * 255), int(b * 255)
        color_code = f"\033[38;2;{r};{g};{b}m"
        colored_token = f"{color_code}{Style.BRIGHT}{text}{Style.RESET_ALL}"
        output_text += ' ' + colored_token
    output_text += '\n'
print(output_text)  
```
Result:
![Screenshot from 2023-08-16 12-59-53](https://github.com/linto-ai/whisper-timestamped/assets/54935496/fade7f5b-d9cc-4f3c-a6c5-026468bcfb8f)

Standard whisper is modified same way as in this [PR ](https://github.com/openai/whisper/pull/1119) . Result:
![Screenshot from 2023-08-16 13-45-14](https://github.com/linto-ai/whisper-timestamped/assets/54935496/635c2c92-5d43-497a-a7f1-143ec7e8350e)
Somehow `whisper` always gives low probs for the end of the sentence. (`whisper` in my examples uses GreedyDecoder). But both models seems giving results far from truth. Maybe you have any clues on that?
PS. Adding` temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0)` makes no difference and adding beam_size = 5, best_of=5 outputs an error since update of BeamSearchDecoder don't have t`oken_probs `among input arguments.

**Comment by Jeronymous** [comment]
@Rtut654
If I understand correctly, you are mentioning differences on the word confidence scores only (the transcription is the same).

> an error since update function of BeamSearchDecoder don't have token_probs among input arguments.

Also this is due to a missing thing in the [PR you mention](https://github.com/openai/whisper/pull/1119) (beam search decoding not implemented).

You are not giving the code you use to print the colorful transcription with whisper. But if I use a similar code as the one you use for whisper-timestamped, it gives the same. I use this code (I added a variable `USE_WHISPER_ORIGINAL` to your code):
```
USE_WHISPER_ORIGINAL = True
if USE_WHISPER_ORIGINAL:
    import whisper
    kwargs_transcribe = dict(word_timestamps=True)
else:
    import whisper_timestamped as whisper
    kwargs_transcribe = dict()
from colorama import init, Style
from typing import List
import colorsys

AUDIO = 'audio/audio_2023-08-16_12-22-08.wav'

audio = whisper.load_audio(AUDIO)
model = whisper.load_model("small", device="cuda:4")
results = whisper.transcribe(model, audio=audio, language='en', **kwargs_transcribe)

output_text = ''
init(autoreset=False)
for segment in results['segments']:    
    for word in segment['words']:
        if USE_WHISPER_ORIGINAL:
            text = word['word'].strip()
            confidence = word['probability']
        else:    
            text = word['text']
            confidence = word['confidence']
        
        r, g, b = colorsys.hsv_to_rgb(confidence * (1/3), 1, 1)
        r, g, b = int(r * 255), int(g * 255), int(b * 255)
        color_code = f"\033[38;2;{r};{g};{b}m"
        colored_token = f"{color_code}{Style.BRIGHT}{text}{Style.RESET_ALL}"
        output_text += ' ' + colored_token
    output_text += '\n'
print(output_text)  
```

> whisper always gives low probs for the end of the sentence

So probably this is a bug in the code you use for colorful transcription with your modified version of whisper.
It's weird to see all the last words of a sentence in red.

> But whisper-timestamped seems giving worse results on the rest of the words.

I haven't understood that. I guess you are expecting too much of Whisper model when you say "incorrect in terms of locating mispronunciation", because the transcription here is correct, it's normal that all word scores are high.

**Comment by Rtut654** [comment]
@Jeronymous Thanks for you fast reply and sorry for the delay! Yes, it was a bug in this PR, if I remember correctly it was always a shift in tokens compare to the real probabilities or something similar. 
I guess you were right about my question - I was expecting too much from this in term of locating mispronunciation, it is far less accurate than I supposed back then.
